chapter_01/final_exercise_01.md

```diff
@@ -20,27 +19,142 @@
+
+
+
 4. **Q:** What is the primary role of the `hadoop-env.sh` configuration file?
+4. **A:** The primary role of the hadoop-env is to set environment variables for hadoop and customize the options of the hadoop processes, like client ops (port and host of the cluster), the size of the Mapreduce heap, the classpath of the hadoop scripts and more, in addition to setting up various jave related options like the Hadoop classname which sets the java class against which the hadoop specific variables that it will run against when the excecution of a process continues
+
 5. **Q:** What purpose does the `core-site.xml` file serve?
+5. **A:** The purpose of the core site is to configure mostly IO related things and also the filesystem and network addresses.
+some examples include: 
+ - [x] hadoop.security.authentication - either simple (no auth) or kerberos, sets whether to use authentication and which type
+ - [x] hadoop.tmp.dir - the base directory where hadoop can store temporary files
+ - [x] hadoop.security.dns.nameserver - the hostname / ip addr of the dns server which the kerberos service nodes use to their own hostname for kerberos login
+ - [x] io.bytes.per.checksum - the number of bytes to use per checksum, must be less then io.file.buffer.size
+ - [x] io.file.buffer.size - sets the buffer size for seqeunce files during read and write opperations (sequence files are binary files containing key-value paris which were created to solve some of the small file problems in hadoop by having multiple files inside one block file)
 
 ### Chapter 2: Hadoop Distributed File System (HDFS)
 
 6. **Q:** What is HDFS?
+6. **A:** HDFS is the Hadoop distributed system which is basically like a regular file system but it is distributed between multiple computers.
+HDFS is built for fault tolorence meaning, it was built with the idea that hardware failures are a norm and happen a lot so hdfs uses replication and redundency to keep the system highly available and fault tolorent.
+HDFS is the primary storage of the Hadoop applications, based on java and which contains data nodes and name nodes, the *name nodes* are the nodes which monitors all the data stored, replica counts and the amount of data nodes which are alive and the FS structure.
+the *data node* is the node that stores the data itself in the form of blocks which have a max size of a preconfigured amount, default is usually 128 MB but sometimes can be 64 or 256 MB by default.
+It is a rack aware system which means that it knows the physical topology of the network and the connected nodes which it uses to enhence fault tolorence by splitting replicas to different racks (an example with 3 replicas and 2 racks, hadoop will store one replica on the same rack with the name node and 2 more replicas on different racks in a different location / switch).
+
+provides High availability using 2 Name nodes with one being active and one being a standby, in case the active node will fail.
```

Comment:
This is called HA mode, but it's not HDFS's default. Please look up what HDFS offers me (Fault tolerance-wise) if HA is truned off.